### ISSUE TITLE
Base rule extension: [func-style] Allow type-annotated variable declaration

### ISSUE THREAD

**Issue by remcohaszing** [typescript-eslint/typescript-eslint#8488]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/?=extension#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/func-style

### Description

If the option is set to `declaration`, an expression would be allowed if the variable has a type annotation.

### Fail

```TypeScript
const foo = function() {
    // …
}
```

### Pass

```TypeScript
function foo() {
    // …
}

const foo: SomeType = function() {
    // …
}
```

### Additional Info

#1236 also suggests to extend this rule, but for other reasons.

**Comment by bradzacher** [comment]
Personally I'm of the opinion that we shouldn't be building more extension rules unless we have really, really strong usecases eg
- there's something fundamentally broken in the rule with TS patterns
- we can greatly improve the rule via type information
- the rule is very, very widely used

I don't think this request fits any of those:
- the rule false-positives specifically if you use a variable assigned function expression with a type annotation
    - this is a very rare combination - function expressions themselves are rare enough in modern JS (most just use an arrow function for the conciseness and better semantics re `this`/`arguments`)
- this change wouldn't require type info
- to my knowledge this rule isn't widely used a quick sourcegraph search [[1]](https://sourcegraph.com/search?q=context%3Aglobal+%2F%5B%22%27%5Dfunc-style%5B%22%27%5D%3A+%5C%5B%3F%28%5B%22%27%5D%28error%7Cwarn%29%5B%22%27%5D%7C2%7C1%29%2F&patternType=keyword&sm=0&filters=%5B%5B%22file%22%2C%22Exclude+node_modules%22%2C%22-file%3A%28%5E%7C%2F%29node_modules%2F%22%5D%5D) shows just 2.6k hits (in comparison `no-unused-vars` hits 24k times [[2]](https://sourcegraph.com/search?q=context:global+/%5B%22%27%5Dno-unused-vars%5B%22%27%5D:+%5C%5B%3F%28%5B%22%27%5D%28error%7Cwarn%29%5B%22%27%5D%7C2%7C1%29/+count:100000&patternType=keyword&sm=0&filters=%5B%5B%22file%22%2C%22Exclude+node_modules%22%2C%22-file%3A%28%5E%7C%2F%29node_modules%2F%22%5D%5D))

I'll leave this open to evaluate community engagement

**Comment by auvred** [comment]
As a `func-style` user, I've already felt the pain of using a function expressions with type annotation :slightly_smiling_face::
 https://github.com/auvred/eslint-config/blob/adddc5a86d57770f58be33ddb0c295836d8f463d/src/plugin/workers/prettier-format.ts#L17-L20

I prefer to use function declarations 90% of the time, but in cases where I need to declare a function that "implements" a particular type, typed function expressions are the only way to do it.

So, if there is some function type expression

```ts
type SharedFunction = (arg1: number, arg2: string) => string
```

I would prefer

```ts
const fn: SharedFunction = (arg1, arg2) => {
  // ...
}
```

Over

```ts
function fn(...[arg1, arg2]: Parameters<SharedFunction>): ReturnType<SharedFunction> {
  // ...
}
```

**Comment by bradzacher** [comment]
My 2c has always been that the declaration style doesn't matter and linting it doesn't meaningfully improve your codebase. If anything I'd consider expressions/arrows safer because they aren't hoisted and thus aren't subject to the same hoisting unsafety as declarations.
But that's neither here nor there for the purposes of this issue.

---

I didn't realise that the rule also enforced arrow funcs vs func decls - so my original statement about the scope isn't right.

---

It's probably worth asking for an upstream enhancement here. This is such a small change that it seems so so so heavy for us to build an entire extension rule around. It could ultimately be covered in an almost 1-line change.

**Comment by remcohaszing** [comment]
I use the following ESLint configuration:

```jsonc
{
  "rules": {
    "func-style": [
      "error",
      "declaration",
      { allowArrowFunctions: true }
    ]
  }
}
```

I personally have a strong preference for function declarations. However, sometimes you want to implement a type. In such cases, you need to use a function expression or an arrow function. Often an arrow function is fine, but sometimes you need the `this` value. In such a case your only option is to use a function expression.

The enhanced rule could use type information to see if the type defines the `this` value. Alternatively it could not use type information, but see if `this` is used. In this case the problem can be solved without type information, but the problem is still specific to TypeScript syntax.

Here’s a practical example: https://github.com/remcohaszing/unified-prettier/blob/v2.0.1/src/unified-prettier.ts#L15-L21

**Comment by bradzacher** [comment]
Correct it is specific to TS syntax.
But for super simple changes like this the ESLint team is sometimes amenable to accepting the change.
For example the rule right now is only 98 lines long (that's with a lot of empty lines).

It would take the addition of a logical expression (`&& !node.parent.typeAnnotation`) in two `if`s to add this feature to their rule (plus tests ofc).
Whereas for us to create an extension rule will require a several dozen lines of code in addition to the documentation and tests and a community migration.

**Comment by bradzacher** [comment]
To be clear - that's why I'm personally against this living in an extension rule.
It's simple but a relatively large chunk of code that we have to maintain going forwards.

Which is why I put a lot of weight on "the rule is widely used".
Our project is now 8 years old and this is the first instance of someone asking for this extension rule.
Which is why I say the rule fails that widely used test -- nobody else has mentioned this in 8 years so it suggests that not a whole lot of people use it with TS code.

**Comment by auvred** [comment]
There are already 4 rules in the ESLint core that use `typeAnnotation` ([GitHub search](https://github.com/search?q=repo%3Aeslint%2Feslint+typeAnnotation+path%3Alib%2F**%2F*&type=code)). So I think it would be fine to request this change on the ESLint side!

**Comment by bradzacher** [comment]
> I've been checking if this was a thing about once a month for 5 years without mentioning it

As a maintainer I interpret that as "the effort to file an issue was much greater than the feature's value to me". Given it only takes a few minutes to file (esp 5y ago when we didn't have issue forms) - that implies the value is super low.

That may sound snarky - but it's not a dig at you. I'm just illustrating how I, as a maintainer, evaluate the value of a request to the userbase. As volunteers we can't entertain all requests (even completely reasonable ones) because all code has maintenance weight as do PRs and follow-up bugs or enhancements. We need to draw the line somewhere so that we don't over-burden ourselves.

**Comment by remcohaszing** [comment]
I didn’t realize the ESLint team is open to supporting `typeAnnotation` in core rules. Solving this in the core rule would be the ideal solution IMO.

The other reason to still want this in `typescript-eslint` is that there’s another feature request to extend `func-style`. If that’s implemented in `typescript-eslint`, then this issue might as well.

**Comment by Josh-Cena** [comment]
Just want to say that I've over the years constantly wanted this functionality too—either as an extension rule, or as a core rule option. I also have the preference of "function declaration, or arrows in callbacks", and in the rare cases where I need both type annotation and `this` (usually Webpack plugins, etc.), I just use a disable comment and move on—but linters shouldn't be built on the premise of "we ban 110% of all known use cases, and you should add disable comments to the last 10% even though we have a clear definition for them"

**Comment by bradzacher** [comment]
If someone wants to ask upstream that would be the first step here. No need for an option, IMO, just a simple implementation of "if the option is 'function declarations' and the variable declaration has a type annotation - ignore it" would probably suffice.

If they don't want to support it (which is a probable response given they have frozen stylistic rules) - I still *personally* don't think we should own support in this project - instead it probably better fits within the "eslint stylistic" project - as it is 100% opinionated code style preference.

I would put support in this project as a 3rd preference.

**Comment by bradzacher** [comment]
I would like to say that there's a big window for inconsistency in allowing this usecase.

What's to stop people from writing code like this:
```ts
type MyFuncType = (arg1: number) => string;
const myFunc: MyFuncType = arg1 => arg1.toString();
```

It passes the rule due to the exception you're adding and it allows a back-door for someone to chose to write expressions and go against the codebase style.

*Personally* I still think that a disable comment and move on is better here for the edge cases of "I need an external function type" as those cases are rarer.

**Comment by Josh-Cena** [comment]
That one is perfectly fine, per any (reasonable) code style. My personal code style is:

1. Use function declarations to bind functions to identifiers. (`function x()`, not `const x = () =>`)
2. If the function needs a type annotation, use an arrow function. (`const x: Fn = () =>`)
3. In the very rare case that the function needs `this` AND needs a type annotation, use a function expression.

I think the case where someone actively tries to use a function expression where unnecessary is rare (unless they don't know about arrow functions...), so all I want is for the rule to not report where necessary. I personally don't mind if the rule allows both styles without nuance.

**Comment by bradzacher** [comment]
> although that means it probably has a much lower rate of usage

As I mentioned above in https://github.com/typescript-eslint/typescript-eslint/issues/8488#issuecomment-1955174480 - the rule already sees low usage.

Which is my worry here and why I don't think this is correct to include in this project.

Based on some quick, naive searches `func-style` is turned on 1/10th as much as `no-unused-vars`. So it's already low on the list of rules people use.

And this change is servicing a subset of that low usage. A subset of a subset is low impact enough that the maintenance weight outweighs the benefit, IMO.

**Comment by kirillgroshkov** [comment]
I just found that this issue 100% describes what we need in our organization.

Can someone point me to an example of how this can be implemented? I would consider forking this rule into our internal codebase and using there (open-sourcing it, of course), I'm ok if it doesn't land in the official repo or stylistic repo, if there's no consensus.

**Comment by kirkwaiblinger** [comment]
@kirillgroshkov 

> Can someone point me to an example of how this can be implemented? 

The source code for eslint's base rule is here https://github.com/eslint/eslint/blob/13d0bd371eb8eb4aa1601c8727212a62ab923d0e/lib/rules/func-style.js. I would use [our playground](https://typescript-eslint.io/play), look at the ESTree view for some code you want to lint, and compare it to the base rule, to get an idea how the rule works, and what you'd need to change for your desired behavior around `typeAnnotation`s.

**Comment by bradzacher** [comment]
5 months later and this request has received 5 community reactions.

Additionally as outlined [in my comment above](https://github.com/typescript-eslint/typescript-eslint/issues/8488#issuecomment-2016975772) the first step was to file an issue with ESLint core and this has not been actioned.

So given both of these - we're taking this as signal that this change is not important broader community and are thus rejecting the proposal. Thanks for filing!
